click function on filters

karmelp opened the pull request:
Miks ta töötab vaid frontendil? kas sellepärast, et kui onclick attribuut on klass filter, siis ta suudab võtta vaid esimese filtriklassi?

karmelp committed:
click function on filters

Miks ta töötab vaid frontendil? kas sellepärast, et kui onclick attribuut on klass filter, siis ta suudab võtta vaid esimese filtriklassi?

karmelp commented:
äkki ma peaks üldse classlisti tegema, et kui vajutada filtril, siis ta toob filtertagi nähtavale, kui uuesti vajutada, siis eemaldab selle

hhabakuk commented:
uuri, mida `document.querySelector...` tagastab. Hetkel sa eeldad kahte asja, mida peab üle vaatama: 
1. `document.querySelector...` tagastab array kõikidest sama klassiga elementidest
2. arrayle saab panna onclick eventi

hhabakuk commented:
>äkki ma peaks üldse classlisti tegema, et kui vajutada filtril, siis ta toob filtertagi nähtavale, kui uuesti vajutada, siis eemaldab selle

See on juba nice-to-have-UX-edasiarendus, tee see kõige viimasena kui soovid

karmelp commented:
> > äkki ma peaks üldse classlisti tegema, et kui vajutada filtril, siis ta toob filtertagi nähtavale, kui uuesti vajutada, siis eemaldab selle
> 
> See on juba nice-to-have-UX-edasiarendus, tee see kõige viimasena kui soovid

aga siis ju jääb nagu praegu frondiga, et saan löputult vajutada frontend nuppu, mis siis teeb filter tagi

karmelp committed:
update

ehhhhk. ma pean köik filter klassiga elemendid tegema nodelistiks ja siis omakorda arrayks?

aga nüüd confusion, et kuidas ma nüü attribuuti kasutada saan, vist ei saagi?

karmelp committed:
turn array into objects

hhabakuk commented on script.js:
```diff
@@ -240,6 +239,50 @@ function getTools(job) {
+
+const filtersNodeListEl = document.querySelectorAll(".filter");
+
+const filters = Array.from(filtersNodeListEl);
```
filtersNodeListEl on array kõikidest filtri elementidest. Kui su eesmärk on igale filtri elemendile onclick külge saada, siis pead selle array üks haaval läbi käima ning susama selle onclicki iga filtri elemendi külge.

`filtersNodeListEl.foreach(filter => console.log('filter', filter))`

hhabakuk replied:
ehk siis kui sul on vaja midagi teha arrayga, siis võid proovida meeles hoida endalt küsida, et kas sa tahad muuta arrayd või sa tahad muuta elementi arrays. Kui sa tahad muuta elementi arrays, siis on alati vaja terve array ükshaaval läbi käia. Sellist in-bulk muutmist javascriptis pole.

hhabakuk replied:
@karmelp see kommentaari ikka veel kehtib.  lihstalt asenda console.log osa funktsiooniga, mis paneks õige onclick funktsiooni külge.

```
function onFilterClick () {
  // toggelda konkreetse filtri css display stiili
}

// pane see funktsioon iga elemendi külge
filtersNodeListEl.foreach(filter => filter.onclick = onFilterClick)
```

karmelp committed:
Update script.js

karmelp committed:
Suutsin onclick funktsiooni saada külge kõikidele filtritele.

karmelp committed:
loo ja eemalda filter

Kuidas ma saan teha nii, et ma saaks if alla panna selle, et kui filtri nupp on deactivated, siis ta eemaldaks selle nupu filtri ribalt?

karmelp committed:
katsetused

karmelp committed:
mingi teine loogika filtri filtriribalt eemaldamiseks

karmelp committed:
Kas sedapidi ei saa? Peab ikka nodelistiga tegema? :D

karmelp committed:
katsedused vol.....

hhabakuk commented:
### Ülesanne: kuva ainult töökuulutusi, mis vastavad valitud filtritele

1. sul on tarvis teha [globaalne muutuja](https://www.w3schools.com/js/js_scope.asp) “let activeFilters = [siia aktiivsed filtrid”]. Globaalseid muutujaid saad defineerida nt viiendal real. Peaasi et nad ei oleks ühegi funktsiooni sees, sest kui nad on funkstiooni sees, siis ainult see funktsioon teab temast

_filtreerimise loogika käiks siis niimoodi, et sul on globaalne muutuja activeFilters. See võib olla nt array. Ehk alguses on const activeFilters = []. Nüüd pead sa hakkama seda arrayd muutma kui klikid mõne filtri aktiivseks. Ehk kui klikid aktiivseks filtrid “Frontend” ja “Junior” peaks activeFilters muutuja sisu olema [“Frontend”, “Junior”]_

2. normaliseeri data struktuur
```
const normalizedJobOffers = data.map((jobOffer) => {
    return {
      id: jobOffer.id,
      company: jobOffer.company,
      logo: jobOffer.logo,
      new: jobOffer.new,
      featured: jobOffer.featured,
      position: jobOffer.position,
      postedAt: jobOffer.postedAt,
      contract: jobOffer.contract,
      location: jobOffer.location,
      // uus!
      // kolm täppi (...), sest tools ja languages on arrayd ja me ei taha, et array sees oleks array.
      filters: [
        jobOffer.level,
        jobOffer.languages,
        ...jobOffer.tools,
        ...jobOffer.languages,
      ],
    };
  });
```

3. filtreeri töökuulutused vastavalt aktiivsetele filtritele
```
  // .filter meetod tagastab ainult need job offerid, mille puhul return tingimus on tõene
  const filteredData = normalizedJobOffers.filter((normalizedJobOffer) => {
    // tingimus on, et iga aktiivne filter (every) peab sisalduma töökuulutuse filtris (includes)
    return activeFilters.every((filter) =>
      normalizedJobOffer.filters.includes(filter)
    );
  });
```

4. kuva ainult filtreeritud kuulutused. 
alumine kood teeb sul äpi katki, sest mujal oled kuvanud jobOffer.languages jne
```
  filteredData.forEach((job) => {
    jobOfferContainer(job);
  });
```

karmelp committed:
filter

karmelp committed:
filtrid tulevad undefined

hhabakuk commented on script.js:
```diff
@@ -234,54 +236,67 @@ function getsvgCircle() {
+function getFilters(job) {
+    const filtersWrapperEl = document.createElement("div");
+    filtersWrapperEl.className = "wrapper"
+    job.filters.forEach((filter) => {
```
Tee siin 

```
console.log('filter argument',filter)
console.log('job.filter', job.filter)
```

karmelp committed:
filter tagid nähtaval

karmelp committed:
Update script.js

karmelp committed:
add and remove elements from activeFilters array

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

hhabakuk commented on script.js:
```diff
@@ -254,4 +228,81 @@ function getsvgCircle() {
+// vali köik filter IDga elemendid
+const filters = document.querySelectorAll("#filter");
+
+filters.forEach(filter => {
```
siin läheks kood selgemaks, kui onfilterclick funktsioon eraldada

```
const onSelectFilter = (event) => {
  // saa teada tekst klikitud nupul
  const filterTitle = event.target.textContent;
  // lisa filter activeFilters arraysse
  activeFilters.push(filterTitle);
  drawDropMenu();
  drawSelectedFilters();
};

filters.forEach((filter) => {
  filter.onclick = (event) => onSelectFilter(event);
});
```

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

hhabakuk commented on script.js:
```diff
@@ -254,4 +228,86 @@ function getsvgCircle() {
+
+// eemalda filter aktiivsete filtrite arrayst
+function removeActiveFilter(activeFilter, filterTagEl) {
+    if (filterTagEl.textContent == activeFilter) {
```
`filterTagEl.textContent == activeFilter` on alati tõene, sest 

`createFilterTag` funkstioonis teed ` filterTagTitleEl.innerHTML = activeFilter`

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
Update script.js

karmelp committed:
added clear all filters btn

karmelp committed:
Update script.js

karmelp committed:
errror

karmelp committed:
done

karmelp committed:
Update script.js
